### PR TITLE
[CALCITE-6607] RexExecutor can throw during evaluation

### PR DESCRIPTION
**Pull request by mihaibudiu**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rex/RexExecutorImpl.java [hunk 1]
```diff
@@ -129,22 +129,20 @@ public static RexExecutable getExecutable(RexBuilder rexBuilder, List<RexNode> e
...
-    String code;
     try {
-      code = compile(rexBuilder, constExps, (list, index, storageType) -> {
+      String code = compile(rexBuilder, constExps, (list, index, storageType) -> {
         throw new UnsupportedOperationException();
       });
+
+      final RexExecutable executable = new RexExecutable(code, constExps);
+      executable.setDataContext(dataContext);
+      executable.reduce(rexBuilder, constExps, reducedValues);
```
Comment (rubenada) [inline core/src/main/java/org/apache/calcite/rex/RexExecutorImpl.java:139]: Is this change really needed? `RexExecutable#reduce` [has already its try-catch](https://github.com/apache/calcite/blob/60e0a3f441a009e55a36cac192253a436bec3f6d/core/src/main/java/org/apache/calcite/rex/RexExecutable.java#L81) and does `reducedValues.addAll(constExps);` in case of a problem

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/rex/RexExecutorImpl.java:139]: Yes, let me attach a stack trace with an exception.

**Comment by mihaibudiu** [comment]
Here is a stack trace which shows that not all exceptions are handled:
```
java.lang.RuntimeException: While compiling [CAST(CAST(MAP('I', 0)):VARIANT NOT NULL):RecordType(INTEGER I, VARCHAR S, INTEGER ARRAY A)]

	at org.apache.calcite.rex.RexExecutable.compile(RexExecutable.java:73)
	at org.apache.calcite.rex.RexExecutable.<init>(RexExecutable.java:53)
	at org.apache.calcite.rex.RexExecutorImpl.reduce(RexExecutorImpl.java:137)
	at org.apache.calcite.rel.rules.ReduceExpressionsRule.reduceExpressionsInternal(ReduceExpressionsRule.java:774)
	at org.apache.calcite.rel.rules.ReduceExpressionsRule.reduceExpressions(ReduceExpressionsRule.java:714)
	at org.apache.calcite.rel.rules.ReduceExpressionsRule$ProjectReduceExpressionsRule.onMatch(ReduceExpressionsRule.java:313)
	at org.apache.calcite.plan.AbstractRelOptPlanner.fireRule(AbstractRelOptPlanner.java:336)
	at org.apache.calcite.plan.hep.HepPlanner.applyRule(HepPlanner.java:556)
	at org.apache.calcite.plan.hep.HepPlanner.applyRules(HepPlanner.java:420)
	at org.apache.calcite.plan.hep.HepPlanner.executeRuleInstance(HepPlanner.java:243)
	at org.apache.calcite.plan.hep.HepInstruction$RuleInstance$State.execute(HepInstruction.java:178)
	at org.apache.calcite.plan.hep.HepPlanner.lambda$executeProgram$0(HepPlanner.java:211)
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:422)
	at org.apache.calcite.plan.hep.HepPlanner.executeProgram(HepPlanner.java:210)
	at org.apache.calcite.plan.hep.HepProgram$State.execute(HepProgram.java:118)
	at org.apache.calcite.plan.hep.HepPlanner.executeProgram(HepPlanner.java:205)
	at org.apache.calcite.plan.hep.HepPlanner.findBestExp(HepPlanner.java:191)
	at org.dbsp.sqlCompiler.compiler.frontend.calciteCompiler.CalciteOptimizer$HepOptimizerStep.optimize(CalciteOptimizer.java:44)
	at org.dbsp.sqlCompiler.compiler.frontend.calciteCompiler.CalciteOptimizer.apply(CalciteOptimizer.java:102)
	at org.dbsp.sqlCompiler.compiler.frontend.calciteCompiler.CalciteCompiler.optimize(CalciteCompiler.java:497)
	at org.dbsp.sqlCompiler.compiler.frontend.calciteCompiler.CalciteCompiler.compile(CalciteCompiler.java:1057)
	at org.dbsp.sqlCompiler.compiler.DBSPCompiler.runAllCompilerStages(DBSPCompiler.java:445)
	at org.dbsp.sqlCompiler.compiler.DBSPCompiler.getFinalCircuit(DBSPCompiler.java:571)
	at org.dbsp.sqlCompiler.compiler.sql.tools.BaseSQLTests.getCircuit(BaseSQLTests.java:324)
	at org.dbsp.sqlCompiler.compiler.sql.tools.BaseSQLTests$CompilerCircuitStream.<init>(BaseSQLTests.java:142)
	at org.dbsp.sqlCompiler.compiler.sql.tools.BaseSQLTests$CompilerCircuitStream.<init>(BaseSQLTests.java:137)
	at org.dbsp.sqlCompiler.compiler.sql.tools.BaseSQLTests.getCCS(BaseSQLTests.java:64)
	at org.dbsp.sqlCompiler.compiler.sql.simple.VariantTests.testQuery(VariantTests.java:72)
	at org.dbsp.sqlCompiler.compiler.sql.simple.VariantTests.testCastMapToStruct(VariantTests.java:432)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:232)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:55)
Caused by: org.codehaus.commons.compiler.CompileException: Line 11, Column 36: Line 11, Column 36: Cannot determine simple type name "Record3_0"
	at org.codehaus.janino.UnitCompiler.getReferenceType(UnitCompiler.java:7144)
	at org.codehaus.janino.UnitCompiler.getReferenceType(UnitCompiler.java:7023)
	at org.codehaus.janino.UnitCompiler.getType2(UnitCompiler.java:6994)
	at org.codehaus.janino.UnitCompiler.access$14900(UnitCompiler.java:240)
	at org.codehaus.janino.UnitCompiler$24.visitReferenceType(UnitCompiler.java:6891)
	at org.codehaus.janino.UnitCompiler$24.visitReferenceType(UnitCompiler.java:6888)
	at org.codehaus.janino.Java$ReferenceType.accept(Java.java:4289)
	at org.codehaus.janino.UnitCompiler.getType(UnitCompiler.java:6888)
	at org.codehaus.janino.UnitCompiler.getType2(UnitCompiler.java:7653)
	at org.codehaus.janino.UnitCompiler.access$15700(UnitCompiler.java:240)
	at org.codehaus.janino.UnitCompiler$25.visitCast(UnitCompiler.java:6918)
	at org.codehaus.janino.UnitCompiler$25.visitCast(UnitCompiler.java:6909)
	at org.codehaus.janino.Java$Cast.accept(Java.java:5283)
	at org.codehaus.janino.UnitCompiler.getType(UnitCompiler.java:6909)
	at org.codehaus.janino.UnitCompiler.getType2(UnitCompiler.java:7432)
	at org.codehaus.janino.UnitCompiler.compileGet2(UnitCompiler.java:5009)
	at org.codehaus.janino.UnitCompiler.access$9000(UnitCompiler.java:240)
	at org.codehaus.janino.UnitCompiler$16.visitConditionalExpression(UnitCompiler.java:4821)
	at org.codehaus.janino.UnitCompiler$16.visitConditionalExpression(UnitCompiler.java:4800)
	at org.codehaus.janino.Java$ConditionalExpression.accept(Java.java:4902)
	at org.codehaus.janino.UnitCompiler.compileGet(UnitCompiler.java:4800)
	at org.codehaus.janino.UnitCompiler.compileGetValue(UnitCompiler.java:5945)
	at org.codehaus.janino.UnitCompiler.access$3800(UnitCompiler.java:240)
	at org.codehaus.janino.UnitCompiler$7.visitRvalue(UnitCompiler.java:2815)
	at org.codehaus.janino.UnitCompiler$7.visitRvalue(UnitCompiler.java:2803)
	at org.codehaus.janino.Java$Rvalue.accept(Java.java:4498)
	at org.codehaus.janino.UnitCompiler.compile(UnitCompiler.java:2803)
	at org.codehaus.janino.UnitCompiler.compileGetValue(UnitCompiler.java:5915)
	at org.codehaus.janino.UnitCompiler.compileGet2(UnitCompiler.java:5888)
	at org.codehaus.janino.UnitCompiler.access$9900(UnitCompiler.java:240)
	at org.codehaus.janino.UnitCompiler$16.visitNewInitializedArray(UnitCompiler.java:4836)
	at org.codehaus.janino.UnitCompiler$16.visitNewInitializedArray(UnitCompiler.java:4800)
	at org.codehaus.janino.Java$NewInitializedArray.accept(Java.java:5770)
	at org.codehaus.janino.UnitCompiler.compileGet(UnitCompiler.java:4800)
	at org.codehaus.janino.UnitCompiler.compileGetValue(UnitCompiler.java:5945)
	at org.codehaus.janino.UnitCompiler.compile2(UnitCompiler.java:2897)
	at org.codehaus.janino.UnitCompiler.access$2800(UnitCompiler.java:240)
	at org.codehaus.janino.UnitCompiler$6.visitReturnStatement(UnitCompiler.java:1594)
	at org.codehaus.janino.UnitCompiler$6.visitReturnStatement(UnitCompiler.java:1579)
	at org.codehaus.janino.Java$ReturnStatement.accept(Java.java:3888)
	at org.codehaus.janino.UnitCompiler.compile(UnitCompiler.java:1579)
	at org.codehaus.janino.UnitCompiler.compileStatements(UnitCompiler.java:1665)
	at org.codehaus.janino.UnitCompiler.compile2(UnitCompiler.java:3713)
	at org.codehaus.janino.UnitCompiler.compile(UnitCompiler.java:3378)
	at org.codehaus.janino.UnitCompiler.compileDeclaredMethods(UnitCompiler.java:1451)
	at org.codehaus.janino.UnitCompiler.compileDeclaredMethods(UnitCompiler.java:1424)
	at org.codehaus.janino.UnitCompiler.compile2(UnitCompiler.java:833)
	at org.codehaus.janino.UnitCompiler.compile2(UnitCompiler.java:446)
	at org.codehaus.janino.UnitCompiler.access$400(UnitCompiler.java:240)
	at org.codehaus.janino.UnitCompiler$3.visitPackageMemberClassDeclaration(UnitCompiler.java:426)
	at org.codehaus.janino.UnitCompiler$3.visitPackageMemberClassDeclaration(UnitCompiler.java:422)
	at org.codehaus.janino.Java$PackageMemberClassDeclaration.accept(Java.java:1688)
	at org.codehaus.janino.UnitCompiler.compile(UnitCompiler.java:422)
	at org.codehaus.janino.UnitCompiler.compile2(UnitCompiler.java:396)
	at org.codehaus.janino.UnitCompiler.access$000(UnitCompiler.java:240)
	at org.codehaus.janino.UnitCompiler$2.visitCompilationUnit(UnitCompiler.java:367)
	at org.codehaus.janino.UnitCompiler$2.visitCompilationUnit(UnitCompiler.java:365)
	at org.codehaus.janino.Java$CompilationUnit.accept(Java.java:371)
	at org.codehaus.janino.UnitCompiler.compileUnit(UnitCompiler.java:365)
	at org.codehaus.janino.SimpleCompiler.cook(SimpleCompiler.java:273)
	at org.codehaus.janino.ClassBodyEvaluator.cook(ClassBodyEvaluator.java:303)
	at org.codehaus.janino.ClassBodyEvaluator.cook2(ClassBodyEvaluator.java:297)
	at org.codehaus.janino.ClassBodyEvaluator.cook(ClassBodyEvaluator.java:273)
	at org.apache.calcite.rex.RexExecutable.compile(RexExecutable.java:64)
	... 57 more
Caused by: org.codehaus.commons.compiler.CompileException: Line 11, Column 36: Cannot determine simple type name "Record3_0"
	at org.codehaus.janino.UnitCompiler.compileError(UnitCompiler.java:13228)
	at org.codehaus.janino.UnitCompiler.getRawReferenceType(UnitCompiler.java:7321)
	at org.codehaus.janino.UnitCompiler.getReferenceType(UnitCompiler.java:7142)
	... 120 more
```

File: core/src/main/java/org/apache/calcite/rex/RexExecutorImpl.java (hunk 1 above)
Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/rex/RexExecutorImpl.java:139]: The crash happens in the constructor for RexExecutable, so the try-catch block in `reduce` is too late.

Comment (rubenada) [reply core/src/main/java/org/apache/calcite/rex/RexExecutorImpl.java:139]: Ok, understood.

File: core/src/main/java/org/apache/calcite/rex/RexExecutorImpl.java [hunk 2]
```diff
@@ -129,22 +129,20 @@ public static RexExecutable getExecutable(RexBuilder rexBuilder, List<RexNode> e
...
     } catch (RuntimeException ex) {
       // Give up on reduction and return expressions unchanged.
       // This effectively moves the error from compile time to runtime.
       // We could give a warning here if there was a mechanism for warnings.
       reducedValues.addAll(constExps);
```
Comment (rubenada) [inline core/src/main/java/org/apache/calcite/rex/RexExecutorImpl.java:144]: Nit: I think your proposal will work fine as of today, but perhaps it'll be safer to do:
```
reducedValues.clear();
reducedValues.addAll(constExps);
```
Just in case in the future we might end up with an exception that leaves `reducedValues` in an intermediate state with some values filled. WDYT?

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/rex/RexExecutorImpl.java:144]: let me make this change

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/rex/RexExecutorImpl.java:151]: I have also added an assertion that on entry the list is empty.

File: core/src/main/java/org/apache/calcite/rex/RexExecutorImpl.java [hunk 3]
```diff
@@ -126,25 +126,30 @@ public static RexExecutable getExecutable(RexBuilder rexBuilder, List<RexNode> e
 
   /**
    * Do constant reduction using generated code.
+   * Returns the reduced expressions in `reducedValues`, which
+   * is supposed to be empty on call.
    */
   @Override public void reduce(RexBuilder rexBuilder, List<RexNode> constExps,
       List<RexNode> reducedValues) {
-    String code;
+    // If this is not empty, the reducedValues.clear() call below is wrong.
```
Comment (NobiGo) [inline core/src/main/java/org/apache/calcite/rex/RexExecutorImpl.java:134]: I don't know Why the java comment is `If this is not empty, the reducedValues.clear() call below is wrong` . Maybe we don't need it, use `assert reducedValues.isEmpty();` is enough.

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/rex/RexExecutorImpl.java:134]: I guess this comment is now redundant with the JavaDoc.
Previously it wasn't obvious that reducedValues must be empty, but I checked all the callers.

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/rex/RexExecutorImpl.java:134]: I have pushed a commit with an improved JavaDoc.

**Comment by rubenada** [comment]
Change LGTM. 
@mihaibudiu  would it be possible to add a unit test for this? (something like the one that provoked the stacktrace that you posted)

**Comment by mihaibudiu** [comment]
I tried a bit to make up a test, but couldn't. The stack trace comes from an application that uses two features: a user-defined type, and the yet unmerged VARIANT data type. I will try to see if I can reverse engineer another test that would fail in the constructor of RexExecutable. Most tests fail in compilation, in the block that already is surrounded by a try-catch.

File: core/src/main/java/org/apache/calcite/rex/RexExecutorImpl.java [hunk 4]
```diff
@@ -126,25 +126,33 @@ public static RexExecutable getExecutable(RexBuilder rexBuilder, List<RexNode> e
...
+      String code = compile(rexBuilder, constExps, (list, index, storageType) -> {
         throw new UnsupportedOperationException();
       });
+
+      final RexExecutable executable = new RexExecutable(code, constExps);
+      executable.setDataContext(dataContext);
+      executable.reduce(rexBuilder, constExps, reducedValues);
     } catch (RuntimeException ex) {
-      // Give up on reduction and return expressions unchanged.
+      // Something went wrong during constant reduction (for example,
```
Comment (caicancai) [inline core/src/main/java/org/apache/calcite/rex/RexExecutorImpl.java:148]: Maybe this part of the javadoc can remove the subject we
```
// An error occurred during constant reduction (for example,
// an attempt to divide by zero).
// Abandon the reduction and return constExps unchanged.
```
or
```
// An error occurred during constant reduction (for instance,
// a division by zero may have been attempted).
// Abandon the reduction process and return constExps unchanged.
```

**Comment by mihaibudiu** [comment]
I know that I should add a test for this fix, but since the only code that I know of that can cause this failure is still not in main, I will merge this without a test. I am quite confident that it is the right solution to the problem.

**Comment by mihaibudiu** [comment]
made a mistake and merged this without squashing the commits. I apologize.
